builder/builder.go

```diff
@@ -720,3 +866,73 @@ func checkRPM(path string) error {
 	}
 	return nil
 }
+
+func parseUint32(s string) (uint32, error) {
+	parsed, err := strconv.ParseUint(s, 10, 32)
+	if err != nil {
+		return 0, errors.Wrapf(err, "error parsing value %q", s)
+	}
+	return uint32(parsed), nil
+}
+
+// createCompressedArchive will use tar and xz to create a compressed
+// file. It does not stream the sources contents, doing all the work
+// in memory before writing the destination file.
+func createCompressedArchive(dst string, srcs ...string) error {
+	err := createCompressedArchiveInternal(dst, srcs...)
+	return errors.Wrapf(err, "couldn't create compressed archive %s", dst)
+}
+
+func createCompressedArchiveInternal(dst string, srcs ...string) error {
+	archive := &bytes.Buffer{}
+	xw, err := swupd.NewExternalWriter(archive, "xz")
+	if err != nil {
+		return err
+	}
+
+	err = archiveFiles(xw, srcs)
+	if err != nil {
+		_ = xw.Close()
+		return err
+	}
+
+	err = xw.Close()
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(dst, archive.Bytes(), 0644)
+}
+
+func archiveFiles(w io.Writer, srcs []string) error {
+	tw := tar.NewWriter(w)
+	for _, src := range srcs {
```

Comment:
Probably need either goto or defer to close the tw in the event of error.

Reply:
We don't need to Close a tar.Writer.

Reply:
...In case of failure.

Reply:
OK.